numpy/core/src/multiarray/descriptor.c

```diff
@@ -2790,17 +2797,35 @@ arraydescr_setstate(PyArray_Descr *self, PyObject *args)
                         return NULL;
                     }
                 }
+#else
+                // PY2 should be able to read PY3 pickles. See gh-2407
+                if (PyString_Check(name)) {
+                    // It is a PY2 string, no transformation is needed
+                    new_name = name;
+                    Py_INCREF(new_name);
+                }
+                else if (PyUnicode_Check(name)) {
+                    // The field names of a structured dtype were pickled in PY3 as unicode strings
+                    // so, to unpickle them in PY2, we need to convert them to PY2 strings
+                    new_name = PyUnicode_AsEncodedString(name, "utf-8", "strict");
```

Comment:
Doesn't this result in an an un-round-trippable field name? I think this encoding needs to match line 2795, else the following will fail:
* Write in py3 with non-ascii field names
* Read in py2
* Write back out in py2
* Read in py3

Choosing `ascii` as an encoding might be safest here, simply because its the intersection of latin1 and utf8, both of which we use elsewhere in compatibility code.

Reply:
@fersarr - thoughts? Can we construct another test that will
- add another bytestring result from the result of step 3
- on python3 load the bytestring and compare it to the original
- on python2 check that the result of the current test matches this second string

Reply:
Hey Hey. Sure, I will give that a go as soon as I am back from holiday!

Reply:
@eric-wieser @mattip I just pushed the new changes, hopefully now we are okay to merge. Thanks for all the help.